beacon-chain/sync/initial-sync/sync_blocks.go

```diff
@@ -0,0 +1,172 @@
+		if strings.HasPrefix(err.Error(), debugError) {
+			log.Debug(strings.TrimPrefix(err.Error(), debugError))
+			return
+		}
+		log.Errorf("Unable to save block: %v", err)
+	}
+}
+
+// processBatchedBlocks processes all the received blocks from
+// the p2p message.
+func (s *InitialSync) processBatchedBlocks(msg p2p.Message) {
+	ctx, span := trace.StartSpan(msg.Ctx, "beacon-chain.sync.initial-sync.processBatchedBlocks")
+	defer span.End()
+	batchedBlockReq.Inc()
+
+	response := msg.Data.(*pb.BatchedBeaconBlockResponse)
+	batchedBlocks := response.BatchedBlocks
+	if len(batchedBlocks) == 0 {
+		// Do not process empty responses.
+		return
+	}
+
+	log.Debug("Processing batched block response")
+	for _, block := range batchedBlocks {
+		s.processBlock(ctx, block)
+	}
+	log.Debug("Finished processing batched blocks")
+}
+
+// requestBatchedBlocks sends out a request for multiple blocks till a
+// specified bound slot number.
+func (s *InitialSync) requestBatchedBlocks(startSlot uint64, endSlot uint64) {
+	ctx, span := trace.StartSpan(context.Background(), "beacon-chain.sync.initial-sync.requestBatchedBlocks")
+	defer span.End()
+	sentBatchedBlockReq.Inc()
+	if startSlot > endSlot {
+		log.Debugf(
+			"Invalid batched request from slot %d to %d",
+			startSlot-params.BeaconConfig().GenesisSlot, endSlot-params.BeaconConfig().GenesisSlot,
+		)
+		return
+	}
+	blockLimit := params.BeaconConfig().BatchBlockLimit
+	if startSlot+blockLimit < endSlot {
+		endSlot = startSlot + blockLimit
+	}
+	log.Debugf(
+		"Requesting batched blocks from slot %d to %d",
+		startSlot-params.BeaconConfig().GenesisSlot, endSlot-params.BeaconConfig().GenesisSlot,
+	)
+	s.p2p.Broadcast(ctx, &pb.BatchedBeaconBlockRequest{
+		StartSlot: startSlot,
+		EndSlot:   endSlot,
+	})
+}
+
+// validateAndSaveNextBlock will validate whether blocks received from the blockfetcher
+// routine can be added to the chain.
+func (s *InitialSync) validateAndSaveNextBlock(ctx context.Context, block *pb.BeaconBlock) error {
+	ctx, span := trace.StartSpan(ctx, "beacon-chain.sync.initial-sync.validateAndSaveNextBlock")
+	defer span.End()
+	if block == nil {
+		return errors.New("received nil block")
+	}
+	root, err := hashutil.HashBeaconBlock(block)
+	if err != nil {
+		return err
+	}
+	if err := s.checkBlockValidity(ctx, block); err != nil {
+		return err
+	}
+	log.Infof("Saving block with root %#x and slot %d for initial sync", root, block.Slot-params.BeaconConfig().GenesisSlot)
+	s.currentSlot = block.Slot
+	s.latestSyncedBlock = block
+
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	// delete block from memory.
+	if _, ok := s.inMemoryBlocks[block.Slot]; ok {
+		delete(s.inMemoryBlocks, block.Slot)
+	}
+	state, err := s.db.State(ctx)
+	if err != nil {
+		return err
+	}
+	if err := s.chainService.VerifyBlockValidity(block, state); err != nil {
+		return err
+	}
+	if err := s.db.SaveBlock(block); err != nil {
+		return err
+	}
+	state, err = s.chainService.ApplyBlockStateTransition(ctx, block, state)
```

Comment:
This does not apply fork choice, it simply runs the state transitions, cleans up block operations, and updates chain head in initial sync